### ISSUE TITLE
[bug] conan.tools.cmake CMakeToolchain doesn't do parallel builds

### ISSUE THREAD

**Issue by TedSmithNXC** [conan-io/conan#10030]
### Environment Details (include every applicable attribute)
  * Operating System+version:  Windows, clang, gcc
  * Compiler+version: all
  * Conan version: 1.40.2
  * Python version:  3.6.9

### Steps to reproduce (Include if Applicable)
use the CMakeToolchain generator and cmake to do a build.  The compiler-appropriate parallel build argument is no longer passed.

I have attached a `conanfile.py`  with lines marked `OLD METHOD` which does parallel builds and `NEW METHOD` which does not.
This is a simple cmake file that builds maincpp.  You will see, for example, -j16 passed when make is invoked on linux for the OLD METHOD but not for the NEW METHOD. (see the logs)

[parallel_build_problem.zip](https://github.com/conan-io/conan/files/7549878/parallel_build_problem.zip)

### Logs (Executed commands with output) (Include/Attach if Applicable)
No parallel build with New `CMakeToolChain` generator (NEW METHOD uncommented from conanfile.py):

```tsmith@tsmith-desktop:~/repos/tito-dev/conan-test$ conan install . -if build -pr:b default -pr:h default
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.py: Installing package
Requirements
Packages

Installing (downloading, building) binaries...
conanfile.py: Generator txt created conanbuildinfo.txt
conanfile.py: Generator 'CMakeToolchain' calling 'generate()'
conanfile.py: Calling generate()
conanfile.py: Aggregating env generators
conanfile.py: Generated conaninfo.txt
conanfile.py: Generated graphinfo
tsmith@tsmith-desktop:~/repos/tito-dev/conan-test$ conan build . -if build -bf build
Using lockfile: '/home/tsmith/repos/tito-dev/conan-test/build/conan.lock'
Using cached profile from lockfile
conanfile.py: Calling build()
conanfile.py: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/tsmith/repos/tito-dev/conan-test/build/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/tsmith/repos/tito-dev/conan-test/build/package" "/home/tsmith/repos/tito-dev/conan-test"

----Running------
> cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/tsmith/repos/tito-dev/conan-test/build/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/tsmith/repos/tito-dev/conan-test/build/package" "/home/tsmith/repos/tito-dev/conan-test"
-----------------
Using Conan toolchain through /home/tsmith/repos/tito-dev/conan-test/build/conan_toolchain.cmake.
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- The CXX compiler identification is GNU 8.4.0
-- The C compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Detecting CXX compiler ABI info
Using Conan toolchain through .
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
Using Conan toolchain through .
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/tsmith/repos/tito-dev/conan-test/build
conanfile.py: CMake command: cmake --build '/home/tsmith/repos/tito-dev/conan-test/build'

----Running------
> cmake --build '/home/tsmith/repos/tito-dev/conan-test/build'
-----------------
/usr/local/bin/cmake -S/home/tsmith/repos/tito-dev/conan-test -B/home/tsmith/repos/tito-dev/conan-test/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/bin/cmake -E cmake_progress_start /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles /home/tsmith/repos/tito-dev/conan-test/build//CMakeFiles/progress.marks
/usr/bin/make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/bin/make  -f CMakeFiles/mytest.dir/build.make CMakeFiles/mytest.dir/depend
make[2]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
cd /home/tsmith/repos/tito-dev/conan-test/build && /usr/local/bin/cmake -E cmake_depends "Unix Makefiles" /home/tsmith/repos/tito-dev/conan-test /home/tsmith/repos/tito-dev/conan-test /home/tsmith/repos/tito-dev/conan-test/build /home/tsmith/repos/tito-dev/conan-test/build /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles/mytest.dir/DependInfo.cmake --color=
Scanning dependencies of target mytest
make[2]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/bin/make  -f CMakeFiles/mytest.dir/build.make CMakeFiles/mytest.dir/build
make[2]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
[ 50%] Building CXX object CMakeFiles/mytest.dir/main.cpp.o
/usr/bin/c++ -D_GLIBCXX_USE_CXX11_ABI=0  -m64 -O3 -DNDEBUG -fPIE -o CMakeFiles/mytest.dir/main.cpp.o -c /home/tsmith/repos/tito-dev/conan-test/main.cpp
[100%] Linking CXX executable mytest
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/mytest.dir/link.txt --verbose=1
/usr/bin/c++ -m64 -O3 -DNDEBUG -m64 CMakeFiles/mytest.dir/main.cpp.o -o mytest 
make[2]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
[100%] Built target mytest
make[1]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/local/bin/cmake -E cmake_progress_start /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles 0
```

Parallel builds with old `cmake` generator (OLD METHOD uncommented from conanfile.py):

```smith@tsmith-desktop:~/repos/tito-dev/conan-test$ conan install . -if build -pr:b default -pr:h default
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.py: Installing package
Requirements
Packages

Installing (downloading, building) binaries...
conanfile.py: Generator txt created conanbuildinfo.txt
conanfile.py: Generator cmake created conanbuildinfo.cmake
conanfile.py: Aggregating env generators
conanfile.py: Generated conaninfo.txt
conanfile.py: Generated graphinfo
tsmith@tsmith-desktop:~/repos/tito-dev/conan-test$ conan build . -if build -bf build
Using lockfile: '/home/tsmith/repos/tito-dev/conan-test/build/conan.lock'
Using cached profile from lockfile
conanfile.py: Calling build()

----Running------
> cd '/home/tsmith/repos/tito-dev/conan-test/build' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="gcc" -DCONAN_COMPILER_VERSION="8" -DCONAN_CXX_FLAGS="-m64" -DCONAN_SHARED_LINKER_FLAGS="-m64" -DCONAN_C_FLAGS="-m64" -DCONAN_LIBCXX="libstdc++" -DCMAKE_INSTALL_PREFIX="/home/tsmith/repos/tito-dev/conan-test/build/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCONAN_CMAKE_POSITION_INDEPENDENT_CODE="ON" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -Wno-dev '/home/tsmith/repos/tito-dev/conan-test'
-----------------
-- The CXX compiler identification is GNU 8.4.0
-- The C compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_INSTALL_BINDIR
    CMAKE_INSTALL_DATAROOTDIR
    CMAKE_INSTALL_INCLUDEDIR
    CMAKE_INSTALL_LIBDIR
    CMAKE_INSTALL_LIBEXECDIR
    CMAKE_INSTALL_OLDINCLUDEDIR
    CMAKE_INSTALL_SBINDIR
    CONAN_CMAKE_POSITION_INDEPENDENT_CODE
    CONAN_COMPILER
    CONAN_COMPILER_VERSION
    CONAN_CXX_FLAGS
    CONAN_C_FLAGS
    CONAN_EXPORTED
    CONAN_IN_LOCAL_CACHE
    CONAN_LIBCXX
    CONAN_SHARED_LINKER_FLAGS


-- Build files have been written to: /home/tsmith/repos/tito-dev/conan-test/build

----Running------
> cmake --build '/home/tsmith/repos/tito-dev/conan-test/build' '--' '-j16'
-----------------
/usr/local/bin/cmake -S/home/tsmith/repos/tito-dev/conan-test -B/home/tsmith/repos/tito-dev/conan-test/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/bin/cmake -E cmake_progress_start /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles /home/tsmith/repos/tito-dev/conan-test/build//CMakeFiles/progress.marks
/usr/bin/make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/bin/make  -f CMakeFiles/mytest.dir/build.make CMakeFiles/mytest.dir/depend
make[2]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
cd /home/tsmith/repos/tito-dev/conan-test/build && /usr/local/bin/cmake -E cmake_depends "Unix Makefiles" /home/tsmith/repos/tito-dev/conan-test /home/tsmith/repos/tito-dev/conan-test /home/tsmith/repos/tito-dev/conan-test/build /home/tsmith/repos/tito-dev/conan-test/build /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles/mytest.dir/DependInfo.cmake --color=
Scanning dependencies of target mytest
make[2]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/bin/make  -f CMakeFiles/mytest.dir/build.make CMakeFiles/mytest.dir/build
make[2]: Entering directory '/home/tsmith/repos/tito-dev/conan-test/build'
[ 50%] Building CXX object CMakeFiles/mytest.dir/main.cpp.o
/usr/bin/c++   -O3 -DNDEBUG -o CMakeFiles/mytest.dir/main.cpp.o -c /home/tsmith/repos/tito-dev/conan-test/main.cpp
[100%] Linking CXX executable mytest
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/mytest.dir/link.txt --verbose=1
/usr/bin/c++ -O3 -DNDEBUG CMakeFiles/mytest.dir/main.cpp.o -o mytest 
make[2]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
[100%] Built target mytest
make[1]: Leaving directory '/home/tsmith/repos/tito-dev/conan-test/build'
/usr/local/bin/cmake -E cmake_progress_start /home/tsmith/repos/tito-dev/conan-test/build/CMakeFiles 0
```

<!--
  Your log content should be related to the bug description, it can be:
  - Conan command output
  - Server output (Artifactory, conan_server)
--
[parallel_build_problem.zip](https://github.com/conan-io/conan/files/7549873/parallel_build_problem.zip)
>

**Comment by memsharded** [comment]
Hi @TedSmithNXC 

I'd say in your example you have the ``CMakeToolchain`` twice, once declared in ``generators=`` and another in the ``generate()`` method, you should use only one of them.

Then, the issue is that the new tools try to decouple the "what" and the "how". The parallel builds is part of the "how", and changing it shouldn't affect at all the final binary. Then, it is defined by configuration items, like ``tools.build:processes``. You can check the available possibilities with ``conan config list``. For example ``tools.cmake.cmaketoolchain:msvc_parallel_compile`` can control de addition of ``/MP`` flags in the CMakeToolchain.

So the functionality to define this seems to be there, but the default has changed and now is not setting parallel builds by default. Let's re-check what would be the best default.

**Comment by kenfred** [comment]
@memsharded 
What does the `parallel` argument mean in the new cmake build helper?

The [doc](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmake.html?highlight=cmake) says:
>parallel: (Optional, Defaulted to True): If True, will append the -jN attribute (/m: in MSBuild) for parallel building being N the value of the tools.microsoft.msbuild:max_cpu_count.

Does this not work?

**Comment by memsharded** [comment]
It should, do you have defined ``tools.microsoft.msbuild:max_cpu_count`` and is still not adding the parallel arguments?

I am still not sure about the purpose of the ``parallel`` argument, for me it reads as some legacy copy&paste, so probably can be dropped. I vaguely recall something like some package parallel build was failing, and the recipe could force ``parallel=False`` to disable it. I think it might be dropped, and fully rely on configuration and defaults for this.

**Comment by TedSmithNXC** [comment]
Oh i see I do have the generator twice.  I'll remove that.

Setting parallel=True in the cmake generator doesn't seem to help.

I tried this:
```
def build(self):
        print("cpus is ", tools.cpu_count())
        cmake = CMake(self, parallel=True)
        cmake.configure()
        cmake.build()
```

I also added a second file to the compile.  

I got this on windows:
```
PS C:\Users\tsmith\repos\cpb> conan install . -if build -pr:b default -pr:h default
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=15
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=15
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

conanfile.py: Installing package
Requirements
Packages

Installing (downloading, building) binaries...
conanfile.py: Generator txt created conanbuildinfo.txt
conanfile.py: Calling generate()
conanfile.py: Aggregating env generators
conanfile.py: Generated conaninfo.txt
conanfile.py: Generated graphinfo
PS C:\Users\tsmith\repos\cpb> conan build . -if build -bf build
Using lockfile: 'C:\Users\tsmith\repos\cpb\build/conan.lock'
Using cached profile from lockfile
conanfile.py: Calling build()
cpus is  12
conanfile.py: CMake command: cmake -G "Visual Studio 15 2017" -DCMAKE_TOOLCHAIN_FILE="C:/Users/tsmith/repos/cpb/build/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/tsmith/repos/cpb/build/package" "C:\Users\tsmith\repos\cpb"
Using Conan toolchain through C:/Users/tsmith/repos/cpb/build/conan_toolchain.cmake.
-- The CXX compiler identification is MSVC 19.16.27040.0
-- The C compiler identification is MSVC 19.16.27040.0
-- The ASM compiler identification is MSVC
-- Found assembler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Detecting CXX compiler ABI info
Using Conan toolchain through .
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
Using Conan toolchain through .
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/tsmith/repos/cpb/build
conanfile.py: CMake command: cmake --build C:\Users\tsmith\repos\cpb\build --config Release
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/tsmith/repos/cpb/CMakeLists.txt
  Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27040 for x64
  Copyright (C) Microsoft Corporation.  All rights reserved.

  cl /c /W1 /WX- /diagnostics:classic /O2 /Ob2 /D _MBCS /D WIN32 /D _WINDOWS /D NDEBUG /D "CMAKE_INTDIR=\"Release\"" /G
  m- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"mytest.dir\Release\\" /Fd"mytest.dir\Release
  \vc141.pdb" /Gd /TP /errorReport:queue C:\Users\tsmith\repos\cpb\main.cpp C:\Users\tsmith\repos\cpb\file2.cpp

  main.cpp
  file2.cpp
  Generating Code...
  mytest.vcxproj -> C:\Users\tsmith\repos\cpb\build\Release\mytest.exe
  Building Custom Rule C:/Users/tsmith/repos/cpb/CMakeLists.txt
PS C:\Users\tsmith\repos\cpb>
```

I also see no change on linux.  

Note I am printing the CPU count which is 12.

**Comment by TedSmithNXC** [comment]
am I setting parallel=True in the wrong place?

**Comment by TedSmithNXC** [comment]
@memsharded   I see now that you commented on `parallel` perhaps doing nothing anymore.   If my tools.cpu_count() reports multicore but I'm still not getting parallel builds, do you have any other ideas as to how to get the build command to include the proper parallel build flags?   I'm not clear on what else I could do to force it.

**Comment by memsharded** [comment]
Yes, apparently you are still not setting the configuration. Please make sure to add in your profile:

```
[conf]
tools.microsoft.msbuild:max_cpu_count=4
```
You can also try with ``tools.build:processes`` (there is still some work to do here, the UX should probably be improved). You can also set those values in command line with ``-c tools.build:processes=4``.

**Comment by kenfred** [comment]
@memsharded

TL;DR: the default should be max parallelism without the need to make any changes to my config or provide cmdline args.

Yes, the UX here leaves much to be desired. The old `cpu_count` mechanism worked well. Having to put the cpu count in the config is very cumbersome because it is machine-specific: it's not a config that we can publish company-wide and have people do `config install`. Having a way to override is useful, but the default should be max parallelism. I don't like having to visit all of our dev systems and docker images to restore the parallelism.

On the subject of docker images, this is especially problematic, since you can dynamically provide the number of CPUs available to the docker image.

I'm also confused as to which variable I should set. I created the following in a `global.conf`:
```
tools.cmake.cmaketoolchain:msvc_parallel_compile=8
tools.microsoft.msbuild:max_cpu_count=8
tools.build:processes=8
tools.ninja:jobs=8
tools.gnu.make:jobs=8
```

All of these are valid for windows and the set seems redundant.

**Comment by TedSmithNXC** [comment]
the change to `global.conf` worked.    How is that file parsed?    Can I use python calls or jinja calls in there?

**Comment by TedSmithNXC** [comment]
The command line and global conf options work well.   I am still of the opinion **changes should be made to have the default be to parallel build with all available cpus as was done before**.

I'm closing this issue because the manual options are sufficient to allow us to accomplish parallel builds.

**Comment by kenfred** [comment]
> The command line and global conf options work well. I am still of the opinion **changes should be made to have the default be to parallel build with all available cpus as was done before**.
> 
> I'm closing this issue because the manual options are sufficient to allow us to accomplish parallel builds.

I'd keep this open for two reasons:
- The CMake helper has a constructor argument `parallel` that is not obeyed. 
- In a CI situation, the number of cores is unknown, so hardcoding it on the command line or in global.conf can't maximize the cores of the CI machine.

**Comment by memsharded** [comment]
Yes, lets keep this issue open, I think there are improvements here to do:

- Change the default to parallel
- Try to normalize and reduce the high number of conf options for this
- Maybe remove the constructor argument (research the reason of this)

**Comment by memsharded** [comment]
It also makes sense to define the machine ``cpu_count`` number in the ``global.conf`` configuration, there could be different approaches:

- Jinja template with function help
- Direct injection of a variable, like PROFILE_DIR is done

cc/ @maikelvdh

**Comment by TedSmithNXC** [comment]
I like the Jinja template idea because it would allow for defining based on a system variable.  We could control the definition file but allow CI or local env's to change from the default by setting the env variable.

**Comment by kenfred** [comment]
> I like the Jinja template idea because it would allow for defining based on a system variable. We could control the definition file but allow CI or local env's to change from the default by setting the env variable.

Of course, this would be a manual way of doing what conan used to do: discover the number of cores on the build machine and default the number of parallel builds accordingly. It feels like a big step backward for us to try and do logic within a jinja template within the conf file when 99.9% of the time, we just want to maximize our cores.

I don't see much use in jinja templating to define build parallelism. For your local dev machine, you either want to max out cores, or fix it to some subset of cores - there is no need for the templating logic. For a CI machine, you want to max out the cores. What other kinds of logic would we expect people to want to do?

**Comment by kenfred** [comment]
> We could control the definition file but allow CI or local env's to change from the default by setting the env variable.

On this point, we can already set config values on the command line, so this back door to a jinja template in the global.conf is unnecessary and kind of a horrible mess.

**Comment by memsharded** [comment]
This is what I am thinking:

- Defaulting to max cores
- Having a way to customize in config, based on the available number of cores, like doing cpu_count/2, or something like that. 

The issue is than when maximizing the cores, it is possible that some machines can become a bit unresponsive, and fail to communicate correctly with the CI system, or be very slow for a developer machine? It seems that there are users out there that don't want to max the cores always, but still want to have some parameterization based on the number of cores available

**Comment by kenfred** [comment]
That sounds great to me!

I agree. There are cases where people don't want max parallelism. In this case, the user can set the number cores in the conf file or on the command line. Something like `cpu_count/2` is not necessary - the user will just set the number of cores explicitly.

Jinja templating like `cpu_count/2` is for when you want to publish a conf file across an organization or within CI. It's definitely the wrong answer across the organization. Your use case of throttling back CI machines is a valid use case. However, I'd suggest that, if people are trying to optimize their CI performance, a blanket config install with a blunt `cpu_count/2` isn't going to be a great choice because it doesn't adapt/scale well to CI system changes.

**Comment by memsharded** [comment]
I am proposing https://github.com/conan-io/conan/pull/10068. 

But I still have one doubt, regarding Visual, if both /m and /MP should be set to max cores by default:

- /m = /maxCpuCount is the number of parallel MSBuild jobs/tasks
- /MP = the number of parallel cl.exe compiled files

The issue is that the effective number of concurrent builds could become /m * /MP, overwhelming the computer (making dev machines for example unresponsive)

Any Visual expert here?

**Comment by memsharded** [comment]
Any Visual Studio expertise here? What would be a good default for /m and /MP parallelism specifier?

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/10068 has been merged, will be in next 1.43 to be released soon.

It will include parallel builds by default, -j for NInja/Make, /MP for VisualStudio based solutions (CMake with Visual Studio generator and MSBuild)
